dipy/denoise/fast_lpca.pyx

```diff
@@ -0,0 +1,255 @@
+                        if k>=m:
+                            x = I[i-m,j,k-m]# q=(1, 0, 1)
+                            update_outer_prod(cov, x, 1)
+                            update_vector(mu, x, 1)
+                    if j>=m:
+                        x = I[i,j-m,k]# q=(0, 1, 0)
+                        update_outer_prod(cov, x, -1)
+                        update_vector(mu, x, -1)
+                        if k>=m:
+                            x = I[i,j-m,k-m]# q=(0, 1, 1)
+                            update_outer_prod(cov, x, 1)
+                            update_vector(mu, x, 1)
+                    if k>=m:
+                        x = I[i,j,k-m]# q=(0, 0, 1)
+                        update_outer_prod(cov, x, -1)
+                        update_vector(mu, x, -1)
+
+                    # Save current integrals for future reference
+                    update_matrix(T[cur_i, j, k], cov, 0)
+                    update_vector(S[cur_i, j, k], mu, 0)
+
+                    # Use integral of current rectangle
+                    # Last corner is (i, j, k) => center is (i-radius, j-radius, k-radius)
+                    if (i>=m-1) and (j>=m-1) and (k>=m-1):
+                        # Divide by the number of samples
+                        for ii in range(ndiff):
+                            mu[ii]/= nsamples
+                            for jj in range(ndiff):
+                                cov[ii,jj] /= nsamples
+                        # Subtract self-outer product of mu from cov
+                        update_outer_prod(cov, mu, -1)
+
+                        #Local covariance ready
+                        # update_matrix(out[i-radius, j-radius, k-radius], cov, 0)
+                        # update_vector(meanout[i-radius, j-radius, k-radius], mu, 0)
+
+                    with gil:
+                        # eigen value decomposition
+                        [d,W] = np.linalg.eigh(cov)
+                        
+                        if((i >= radius and i < n0-radius) and (j >= radius and j < n1-radius) and (k >= radius and k < n2-radius)):
+                        # make the projection and reconstruct
+                            temp = I[i - radius: i + radius + 1,
+                                       j - radius: j + radius + 1,
+                                       k - radius: k + radius + 1, :]
+                            temp = np.array(temp)
+                            
+                            X = temp.reshape(m * m * m, ndiff)
+                            X = X - np.array([mu, ] * X.shape[0],
+                                             dtype=np.float64)
+
+                            d[d < tou[i, j, k, :]] = 0
+                            # Y = X.dot(W)
+                            W_hat = W * 0
+                            W_hat[:, d > 0] = W[:, d > 0]
+
+                            Y = X.dot(W_hat)
+                            X_est = Y.dot(np.transpose(W_hat))
```

Comment:
This is what we need to do signal by signal: first precompute `P = W_hat.dot(np.transpose(W_hat))`, and then iterate over all voxels within the window and multiply by P